protocol/protocol.md

```diff
@@ -0,0 +1,162 @@
+
+**Phase 2: KAS Processing & Verification** *(Executed by KAS)*
+
+7.  **Authenticate Client:** Verify the client's authentication credentials. If invalid, return an authentication error.
+8.  **Decrypt DEK Share:** Use the KAS's *private* RSA key (corresponding to the public key used for wrapping, identified by `kid`) to decrypt the `wrappedKey` provided in the request, yielding the plaintext DEK share.
+9.  **Validate Policy Binding:**
+    *   Recalculate the HMAC: `HMAC(DEK share, policy string from request)` using the algorithm specified in the request's `policyBinding.alg`.
+    *   Compare the recalculated HMAC hash with the `policyBinding.hash` provided in the request.
+    *   If they do not match, return a policy binding error (indicates tampering or mismatch).
+10. **(Optional) Decrypt Metadata:** If `encryptedMetadata` was provided, decrypt it using the plaintext DEK share. This metadata might inform policy decisions or logging.
+11. **Perform Authorization Check:**
+    *   Retrieve the requesting client's validated attributes/entitlements (based on their authenticated identity).
+    *   Parse the `policy` string from the request to get the required `dataAttributes` and `dissem` list.
+    *   Evaluate the policy rules (potentially retrieving external Attribute Definitions) against the client's attributes.
+    *   Check if the client is in the `dissem` list (if applicable).
+    *   If authorization fails (policy requirements not met), return an authorization error.
+12. **Rewrap DEK Share:** If all checks pass, encrypt the plaintext DEK share using the client's public key provided in the request (e.g., using RSAES-OAEP if the client key is RSA). Base64 encode the result.
+13. **Send Response:** Return a success response containing the Base64 encoded, rewrapped DEK share.
+
+**Phase 3: Client Decryption** *(Executed by SDK)*
+
+14. **Receive Response:** Get the response from the KAS. Check for errors (authentication, binding, authorization failures).
+15. **Decrypt DEK Share:** If the request was successful, use the client's *private* key to decrypt the rewrapped DEK share received from the KAS.
+16. **(If Key Splitting) Reconstruct DEK:** If multiple shares were required (`sid` was used), combine the decrypted shares (e.g., via XOR) to reconstruct the full plaintext DEK.
+17. **Decrypt Payload:** Use the plaintext DEK and the parameters from `encryptionInformation.method` (IV) to decrypt the TDF payload. During decryption (especially with AES-GCM or streaming), simultaneously verify the integrity of each segment using the `hash` from the [Segment Object](./opentdf/segment.md) and finally verify the `rootSignature` from [`integrityInformation`](./opentdf/integrity_information.md). If any integrity check fails, abort decryption and report an error.
+18. **Securely Discard DEK:** Once the payload is decrypted or decryption fails, securely erase the plaintext DEK and any intermediate shares from memory.
+
+## Error Handling
```

Comment:
we had an adr around this a while back... not sure if we'd wanna include that here, maybe not since we havent implemented against it much . https://github.com/virtru-corp/adr/pull/102

Reply:
I might forgo linking to this ADR for a couple reasons.

1. It's not public
2. The error codes themselves aren't defined by the spec - we only define them for the purposes of our "reference implementation".

What do you think? Should be define the error codes formally here?

Reply:
ahh that is a great point actually. agreed.